### PR TITLE
EC Zoo pull request for Shuubham Ojha

### PR DESCRIPTION
**Pull request by shuubham4**
Pull request for kerdock code.

- Shuubham

### CODE REVIEW

File: kerdock.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# code id, physical, logical are all lower case
+code_id: kerdock
+physical: bits
+logical: bits
+
+name: 'kerdock'
+introduced: '\url{https://doi.org/10.1016/S0019-9958(72)90376-2}'
+
+description: 'Kerdock code is a \([[2^m, 2^{2m}, 2^{m-1} - 2^{frac{m}{2}-1}]]\) binary code where \(m\) is even. The size of code book is \(|2^{m-1}||R(1,m)|\) where \(R(1,m)\)
+is the corresponding first order Reed Muller code. The code is non linear and can exist for perimeters for which there is no linear code (See Nordstrom-Robinson). The relative
```
Comment (AlexanderBarg) [inline kerdock.yml:15]: Replace "The code is non linear and can exist for perimeters for which there is no linear code (See Nordstrom-Robinson)." 
with "The code is nonlinear, and its size is twice the size of the largest possible linear code with the same length and distance".

File: kerdock.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#protection: ''
+
+features:
+
+  rate: 'The transmission rate is \(\frac{2m}{2^m}\) which tends to 0 as \(m\) becomes large, hence these codes are asymptotically poor.'
+
+  encoders:
+    - 'Boolean functions are of the form \(Q(X) + l(X) + b\)' where \(Q(X) is \sum_{1 \leq i < j \leq m} a_{ij}X_{i}X_{j}\), \(l(X)\) is linear in \(m\) variables and \(b\) is 
+       in \(\mathbb{F}_{2}\). 
+    - 'Codewords can be considered as evaluations of these functions over \(\mathbb{F}_{2}^{m}\).'
```
Comment (AlexanderBarg) [inline kerdock.yml:27]: "can be considered" -> "are formed"

File: kerdock.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    - 'Soft decision decoding involves extending the Fast Hadamard Transform decoding algorithm for the binary first-order Reed-Muller code to Kerdock code.'
+    - 'Complexity of soft decision decoding algorithm: \(4^m\) multiplications and \(m4^m\) additions \cite{arXiv:math/0207208}.'
+    
+
+realizations:
+  - 'Kerdock codes find application in digital fingerprinting.'
+  - 't - collision secure schemes can be designed to detect a pirated copy once t users have colluded \cite{doi:10.1109/TIT.2004.838106}.'
+
+notes:
+  - 'A kerdock set is a set of symmetric binary matrices with zero diagonal entries with the property that difference of distinct matrices in the set have full rank.'
```
Comment (AlexanderBarg) [inline kerdock.yml:40]: Kerdock set

File: kerdock.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+features:
+
+  rate: 'The transmission rate is \(\frac{2m}{2^m}\) which tends to 0 as \(m\) becomes large, hence these codes are asymptotically poor.'
+
+  encoders:
+    - 'Boolean functions are of the form \(Q(X) + l(X) + b\)' where \(Q(X) is \sum_{1 \leq i < j \leq m} a_{ij}X_{i}X_{j}\), \(l(X)\) is linear in \(m\) variables and \(b\) is 
+       in \(\mathbb{F}_{2}\). 
+    - 'Codewords can be considered as evaluations of these functions over \(\mathbb{F}_{2}^{m}\).'
+    - 'The matrix \(A = [a_{ij}]\) constitute the Kerdock set \(K\) (See notes).'
```
Comment (AlexanderBarg) [inline kerdock.yml:28]: "matrix" -> "matrices"

File: preparata.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# code id, physical, logical are all lower case
+code_id: preparata
+physical: bits
+logical: bits
+
+name: 'preparata'
+introduced: '\url{https://doi.org/10.1007/0-387-34799-2_39}'
+
+description: 'Preparata code is a \([[2^{m+1}-1, 2^{m+1}-2m-2, 5]]\) binary code where \(m\) is odd. The code is non linear and can exist for perimeters 
+for which there is no linear code. Extended preparata codes can be thought of as characteristic function of the tuple (X,Y) where X, Y belongs to \(\mathbb{F}_{2^m}\) along with the
```
Comment (AlexanderBarg) [inline preparata.yml:15]: Replace "The code is non linear and can exist for perimeters for which there is no linear code."
with "The code is nonlinear, and its size is twice the size of the largest possible linear code with the same length and distance".

Comment (AlexanderBarg) [inline preparata.yml:15]: "Extended preparata codes can be thought of as characteristic function of the tuple (X,Y) where X, Y belongs to \(\mathbb{F}_{2^m}\) along with the following properties"

This sentence is mathematically incorrect, and grammar is flawed.
with

"Codewords of the extended Preparata code can be

File: preparata.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+notes:
+  - 'Preparate codes are uniformly packed and nearly perfect which implies that for any word \(textbf{u}\)' and (preparata) codebook C with \(d(\textbf{u}, C) > 2\), we have
+     that \(u\) has a distance 2 or 3 to exactly floor((2^{m+1}-1)/3) codewords.'  
+  - 'Preparata codes contain twice as many code words as the double-error-correcting BCH codes of the same length, which is the largest number of code words
+     possible for given length and distance. Thus, they are optimal\cite{https://doi.org/10.1016/S0019-9958(68)90874-7}.'
+  - 'Preparata codes are linear over Z4 with the Lee distance.'    
+
+relations:
+  parents:
+    - code_id: bch
```
Comment (AlexanderBarg) [inline preparata.yml:40]: BCH is a relative not a parent

File: preparata.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    - 'Trellis diagrams employed for soft decision decoding. '
+
+notes:
+  - 'Preparate codes are uniformly packed and nearly perfect which implies that for any word \(textbf{u}\)' and (preparata) codebook C with \(d(\textbf{u}, C) > 2\), we have
+     that \(u\) has a distance 2 or 3 to exactly floor((2^{m+1}-1)/3) codewords.'  
+  - 'Preparata codes contain twice as many code words as the double-error-correcting BCH codes of the same length, which is the largest number of code words
+     possible for given length and distance. Thus, they are optimal\cite{https://doi.org/10.1016/S0019-9958(68)90874-7}.'
+  - 'Preparata codes are linear over Z4 with the Lee distance.'    
+
+relations:
```
Comment (AlexanderBarg) [inline preparata.yml:38]: Kerdock codes, 2nd order Reed-Muller codes

File: preparata.yml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  parents:
+    - code_id: bch
+    - detail: 'Preparata codes contain twice as many codewords as the corresponding extended primitive double-error-correcting BCH codes.'
+
+  cousins:
```
Comment (AlexanderBarg) [inline preparata.yml:43]: Kerdock codes, 2nd order Reed-Muller codes

File: kerdock.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+##       https://github.com/errorcorrectionzoo       ##
+#######################################################
+
+# code id, physical, logical are all lower case
+code_id: kerdock
+physical: bits
+logical: bits
+
+name: 'kerdock'
+introduced: '\url{https://doi.org/10.1016/S0019-9958(72)90376-2}'
```
Comment (valbert4) [inline kerdock.yml:12]: \cite{doi:10.1016/S0019-9958(72)90376-2} (please fix any refs elsewhere to conform to format)

File: kerdock.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+realizations:
+  - 'Kerdock codes find application in digital fingerprinting.'
+  - 't - collision secure schemes can be designed to detect a pirated copy once t users have colluded \cite{doi:10.1109/TIT.2004.838106}.'
+
+notes:
+  - 'A kerdock set is a set of symmetric binary matrices with zero diagonal entries with the property that difference of distinct matrices in the set have full rank.'
+  - 'Kerdock codes can be seen as an extended cyclic code over Z4 (quaternary code), where the appended bit is a zero sum check \cite{arXiv:math/0207208}.'
+
+relations:
+  parents:
```
Comment (valbert4) [inline kerdock.yml:44]: 2 parents:
bits_into_bits
quaternary_over_z4 with detail mentioning how these nonlinear binary codes are actually linear codes over Z4 via the Gray map
make reed_muller a cousin (kerdock is nonlinear and reedmuller is a child of linear)

File: kerdock.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    - code_id: reed_muller
+      detail: 'Kerdock code is a subcode of second order Reed Muller Code. It consists of a number of cosets of RM(2,m) created by quotienting with RM(1,m).'
+  cousins:
+    - code_id: nordstrom_robinson
+      detail: 'Specializes Kerdock code for \(m = 4\). Consequently, the Kerdock set is a set of 8 matrices in \(\mathbb(F)_{2}^{4X4}\). The resulting code is binary with 
+      parameters [16, 256, 6].'
+
```
Comment (valbert4) [inline kerdock.yml:51]: Add some more cousins (e.g., preparata) discussed on pg. 386 of Huffman Kim Sole.

Comment (valbert4) [reply kerdock.yml:51]: Remove nordstrom-robinson; it is a child and will be mentioned in that code's entry

File: kerdock.yml [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  - 't - collision secure schemes can be designed to detect a pirated copy once t users have colluded \cite{doi:10.1109/TIT.2004.838106}.'
+
+notes:
+  - 'A kerdock set is a set of symmetric binary matrices with zero diagonal entries with the property that difference of distinct matrices in the set have full rank.'
+  - 'Kerdock codes can be seen as an extended cyclic code over Z4 (quaternary code), where the appended bit is a zero sum check \cite{arXiv:math/0207208}.'
+
+relations:
+  parents:
+    - code_id: reed_muller
+      detail: 'Kerdock code is a subcode of second order Reed Muller Code. It consists of a number of cosets of RM(2,m) created by quotienting with RM(1,m).'
```
Comment (valbert4) [inline kerdock.yml:46]: Don't forget \(...\) for any math like RM().

File: preparata.yml (hunk 6 above)
Comment (valbert4) [reply preparata.yml:40]: Move to cousin. Parents are bits_into_bits and quaternary_over_z4

File: preparata.yml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+# code id, physical, logical are all lower case
+code_id: preparata
+physical: bits
+logical: bits
+
+name: 'preparata'
+introduced: '\url{https://doi.org/10.1007/0-387-34799-2_39}'
+
+description: 'Preparata code is a \([[2^{m+1}-1, 2^{m+1}-2m-2, 5]]\) binary code where \(m\) is odd. The code is non linear and can exist for perimeters
```
Comment (valbert4) [inline preparata.yml:14]: Nonlinear binary code notation is (n,K,d); same for Kerdock

File: kerdock.yml [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+minimum distance tends to \(\frac{1}{2}\) as \(m\) becomes large.'
+
+#protection: ''
+
+features:
+
+  rate: 'The transmission rate is \(\frac{2m}{2^m}\) which tends to 0 as \(m\) becomes large, hence these codes are asymptotically poor.'
+
+  encoders:
+    - 'Boolean functions are of the form \(Q(X) + l(X) + b\)' where \(Q(X) is \sum_{1 \leq i < j \leq m} a_{ij}X_{i}X_{j}\), \(l(X)\) is linear in \(m\) variables and \(b\) is
```
Comment (valbert4) [inline kerdock.yml:25]: Please describe the context of these boolean functions. Are they helping determine the codewords? If so, place into description, which needs a way to define the code anyway.

File: kerdock.yml [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  rate: 'The transmission rate is \(\frac{2m}{2^m}\) which tends to 0 as \(m\) becomes large, hence these codes are asymptotically poor.'
+
+  encoders:
+    - 'Boolean functions are of the form \(Q(X) + l(X) + b\)' where \(Q(X) is \sum_{1 \leq i < j \leq m} a_{ij}X_{i}X_{j}\), \(l(X)\) is linear in \(m\) variables and \(b\) is 
+       in \(\mathbb{F}_{2}\). 
+    - 'Codewords can be considered as evaluations of these functions over \(\mathbb{F}_{2}^{m}\).'
+    - 'The matrix \(A = [a_{ij}]\) constitute the Kerdock set \(K\) (See notes).' 
+
+  decoders:
+    - 'Soft decision decoding involves extending the Fast Hadamard Transform decoding algorithm for the binary first-order Reed-Muller code to Kerdock code.'
```
Comment (valbert4) [inline kerdock.yml:31]: ref?

File: kerdock.yml [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    - 'Complexity of soft decision decoding algorithm: \(4^m\) multiplications and \(m4^m\) additions \cite{arXiv:math/0207208}.'
+    
+
+realizations:
+  - 'Kerdock codes find application in digital fingerprinting.'
```
Comment (valbert4) [inline kerdock.yml:36]: Would be cool if u can find a ref

File: kerdock.yml [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  - 't - collision secure schemes can be designed to detect a pirated copy once t users have colluded \cite{doi:10.1109/TIT.2004.838106}.'
+
+notes:
+  - 'A kerdock set is a set of symmetric binary matrices with zero diagonal entries with the property that difference of distinct matrices in the set have full rank.'
+  - 'Kerdock codes can be seen as an extended cyclic code over Z4 (quaternary code), where the appended bit is a zero sum check \cite{arXiv:math/0207208}.'
```
Comment (valbert4) [inline kerdock.yml:41]: This note can be a detail statement for the quaternary_over_z4 parent

File: preparata.yml [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+name: 'preparata'
+introduced: '\url{https://doi.org/10.1007/0-387-34799-2_39}'
+
+description: 'Preparata code is a \([[2^{m+1}-1, 2^{m+1}-2m-2, 5]]\) binary code where \(m\) is odd. The code is non linear and can exist for perimeters 
+for which there is no linear code. Extended preparata codes can be thought of as characteristic function of the tuple (X,Y) where X, Y belongs to \(\mathbb{F}_{2^m}\) along with the
+following properties:- (i) \(|X| = |Y| = 2^{m}\) (ii) \(\sum_{x \in X} x = \sum_{y \in Y} y \) (iii) \(\sum_{x \in X} x^{3} + (\sum_{x \in X} x)^{3}  = \sum_{y \in Y} y^{3} \). Preparata
+codes are then obtained by leaving out the coordinate in the zeroeth positions in the first half.'
+
+protection: 'Preparata codes are a class of non linear double error correcting codes.'
```
Comment (valbert4) [inline preparata.yml:19]: What is a "double error"? Elaborate on protection please.

File: preparata.yml [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+
+features:
+
+  rate: 'The transmission rate is \(\frac{2^{m+1}-2m-2}{2^{m+1}-1}\).'
+
+  encoders:
+    - ''
+
+  decoders:
+    - 'Trellis diagrams employed for soft decision decoding.'
```
Comment (valbert4) [inline preparata.yml:29]: ref?

File: preparata.yml [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+notes:
+  - 'Preparate codes are uniformly packed and nearly perfect which implies that for any word \(textbf{u}\)' and (preparata) codebook C with \(d(\textbf{u}, C) > 2\), we have
+     that \(u\) has a distance 2 or 3 to exactly floor((2^{m+1}-1)/3) codewords.'  
+  - 'Preparata codes contain twice as many code words as the double-error-correcting BCH codes of the same length, which is the largest number of code words
```
Comment (valbert4) [inline preparata.yml:34]: Move this to detail statement for cousin: bch

File: preparata.yml [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+     possible for given length and distance. Thus, they are optimal\cite{https://doi.org/10.1016/S0019-9958(68)90874-7}.'
+  - 'Preparata codes are linear over Z4 with the Lee distance.'
```
Comment (valbert4) [inline preparata.yml:36]: Move this to detail statement for cousin: quaternary_over_z4

File: preparata.yml [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+  rate: 'The transmission rate is \(\frac{2^{m+1}-2m-2}{2^{m+1}-1}\).'
+
+  encoders:
+    - ''
+
+  decoders:
+    - 'Trellis diagrams employed for soft decision decoding.'
+
+notes:
+  - 'Preparate codes are uniformly packed and nearly perfect which implies that for any word \(textbf{u}\)' and (preparata) codebook C with \(d(\textbf{u}, C) > 2\), we have
```
Comment (valbert4) [inline preparata.yml:32]: Move to description

**Pull request closed without merging**